File: elements/4_linkedlist/LinkedList.java
/*
singley linked list with sentinel head
*/
public class LinkedList<T> {
	Node<T> head = new Node<T>(null);

	/*
	search, O(n)
	*/
	public Node<T> search(T data) {
		Node<T> node = head;
		while(node != null) {
			if(data.equals(node.data)) return node;
			node = node.next;
		}
		return null;
	} 

	/*

	*/
	public void add(T data) {
		Node<T> node = new Node<T>(data);
		node.next = head.next; head.next = node;
	}

	public void delete(T data) {
		Node<T> node = head;
		while(node.next != null) {
			if(data.equals(node.next.data)) { node.next = node.next.next; return; }
			node = node.next;
		}
	}

	public void print() {
		Node<T> node = head;
		while(node != null) { 
			if(node.data != null) System.out.print(node.data + ",");
			node = node.next;
		}
		System.out.println();
	}

	/*
	recursively get to the last node and add from there
	tail recrusion using 2 ref [ ]
	TODO [ ] why this is not working?
	*/
	void reverse(Node<T> node) {
		if(node.next == null) { head.next = node; return; }
		reverse(node.next);

		if(node.data == null) node.next.next = null; 
		else node.next.next = node;
	}

	public void reverse2() {
		reverse(head);
	}
	
	/*
	iterative

	*/
	public void reverse() {
		Node<T> prev = null, cur = head.next, next = null;
		while(cur != null) {
			next = cur.next; // save
			cur.next = prev; // reverse
			prev = cur; cur = next; // iterate
		}
		head.next = prev;
	}

	public static void main(String[] args) {
		LinkedList<String> names = new LinkedList<String>();
		names.add("brian");
		names.add("ben");
		names.add("brian jr");
		names.add("susana");
		names.print();

		//System.out.println(names.search("brian").data);
		names.reverse2();
		names.print();
		
	// 	names.delete("brian"); // first
	// 	names.print();

	// 	names.delete("susana"); // last
	// 	names.print();
	}
}

File: elements/4_linkedlist/LinkedList.cpp
#include <memory>
#include <iostream>
using namespace std;

template<typename T>
struct Node {
	T data;
	shared_ptr<Node<T>> next; 
	Node<T>(T d, shared_ptr<Node<T>> n) { data = d; next = n; }
};

class LinkedList {
	shared_ptr<Node<int>> head; // dummy head

public:
	LinkedList() { head = make_shared<Node<int>>(-1, nullptr); }

	shared_ptr<Node<int>> find(int key) {
		auto cur = head->next;
		while(cur && cur->data != key) cur = cur->next;
		return cur;
	}

	void insert(int key) {
		head->next = make_shared<Node<int>>(key, head->next); 
	}

	void remove(int key) {
		auto cur = head->next, prev = head;
		while(cur && cur->data != key) { prev = cur; cur = cur->next; }
		prev->next = cur->next;
	}

	void print() {
		auto cur = head->next;
		while(cur) { cout << cur->data << " "; cur = cur->next; }
		cout << endl;
	}
};

/*

*/
Node<int>* common(Node<int>* l1, Node<int>* l2) {
	return null;
}

int main() {
	LinkedList list;
	list.insert(1);
	list.insert(2);
	list.insert(3);
	list.print();

	auto node = list.find(1);
	if(node) cout << node->data << endl;

	node = list.find(4);
	if(node) cout << node->data << endl;

	list.remove(2);
	list.print();
	return 0;
}